### PR TITLE
Feature/webrtc qos dscp

### PR DESCRIPTION
**Pull request by slabajo**
## What is the current behavior you want to change?
Media Communication in Internet is subject to best effort service type and has to fight for shared resources in intermediate communication infrastructure. However there exist the DSCP/ToS mechanism that allows an easy tagging and classification of traffic that allows intermediate equipment to manage traffic differentiated and thus providing QoS networks: https://en.wikipedia.org/wiki/Differentiated_services

On RFC8837 there is a proposal to network packet markings using DSCP for WebRTC services: https://datatracker.ietf.org/doc/html/rfc8837#section-5

On modern network equipoment that honors the DSCP packet tagging, WebRTC media traffic can get better latency, priority and less packet drop probablitiy, thus enhancing overall communication quality.

## What is the new behavior provided by this change?
This change implements the ability to tag WebRTC outgoing traffic from Kurento with DSCP values according to https://datatracker.ietf.org/doc/html/rfc8837#section-5
It implements an optional parameter in WebRTCEndpoint constructor, that if present will mark all outgoing network packets with the specified DSCP value.
It also implements an additional configuration parameter in the WebRtcEndpoint.conf.ini configuration file that allows to specify the default DSCP value to apply to all WebRTCEndpoints when no value is specified in constructor.

Of course, this change only applies to outgoing traffic from Kurento media server, to build a complete communication, traffic from browser (or the other peer) should also be tagged accordingly. For browsers the WebRTC Priority Control API (https://www.w3.org/TR/webrtc-priority/) should be used if available.

It must be noted that DSCP markings are applied at the WebRTCEndpoint level so it applies to all streams managed by the same WebRTCEndpoint (audio, video, datachannel).

## How has this been tested.
It has been tested using one of the Kurento tutorials, and taking network dumps of the packets exchanged to verify that  DSCP/ToS markings have been applied corresponding to the parameter in constructor and/or configuration file.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature / enhancement (non-breaking change which improves the project)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] My change requires a change to the documentation
- [ ] My change requires a change in other repository 

## Checklist

- [X] I have read the [Contribution Guidelines](https://github.com/Kurento/.github/blob/master/CONTRIBUTING.md)
- [X] I have added an explanation of what the changes do and why they should be included
- [X] I have written new tests for the changes, as applicable, and have successfully run them locally

### CODE REVIEW

File: src/server/config/WebRtcEndpoint.conf.ini [hunk 1]
```diff
@@ -140,3 +140,38 @@
 ;; <iceTcp> is either 1 (ON) or 0 (OFF). Default: 1 (ON).
 ;;
 ;iceTcp=1
+
+;; Enable DSCP tagging for QoS management
```
Comment (j1elo) [inline src/server/config/WebRtcEndpoint.conf.ini:144]: Minor typo: End summary sentence with period.

File: src/server/config/WebRtcEndpoint.conf.ini [hunk 2]
```diff
@@ -140,3 +140,38 @@
...
+;; WebRTCEndpoints that have this property set to a value different from NO_VALUE
+;; will have its output network packets tagged with the corresponding DSCP value
```
Comment (j1elo) [inline src/server/config/WebRtcEndpoint.conf.ini:146]: Period.

File: src/server/config/WebRtcEndpoint.conf.ini [hunk 3]
```diff
@@ -140,3 +140,38 @@
...
+;; This in turn may make intermediate routers to adapt ist traffic handling for those packets 
+;; according to the DSCP value set.
+;; Recommended values for WebRTc traffic are indicated in section 5 of RFC 8837 https://datatracker.ietf.org/doc/html/rfc8837#section-5
```
Comment (j1elo) [inline src/server/config/WebRtcEndpoint.conf.ini:149]: WebRTc -> WebRTC

File: src/server/interface/elements.WebRtcEndpoint.kmd.json [hunk 4]
```diff
@@ -505,6 +505,23 @@
               "type": "CertificateKeyType",
               "optional": true,
               "defaultValue": "RSA"
+            },
+            {
+              "name": "qosDscp",
+              "doc": "DSCP value to be used in network traffic sent from this endpoint
+              <p>
+              If this parameter is present with a value different to NO_VALUE, all traffic sent from this WebRTCEndpoint will be marked with the correspondent
+              DSCP value according to the DiffServ architecture. This may provide beter handling of network traffic according to its characteristics associated to
```
Comment (j1elo) [inline src/server/interface/elements.WebRtcEndpoint.kmd.json:514]: beter -> better

File: src/server/interface/elements.WebRtcEndpoint.kmd.json [hunk 5]
```diff
@@ -505,6 +505,23 @@
...
               "optional": true,
               "defaultValue": "RSA"
+            },
+            {
+              "name": "qosDscp",
+              "doc": "DSCP value to be used in network traffic sent from this endpoint
+              <p>
+              If this parameter is present with a value different to NO_VALUE, all traffic sent from this WebRTCEndpoint will be marked with the correspondent
+              DSCP value according to the DiffServ architecture. This may provide beter handling of network traffic according to its characteristics associated to 
+              its DSCP value. This better handling include priority in forwarding traffic and probablity of packet drop in case of network congestion.
```
Comment (j1elo) [inline src/server/interface/elements.WebRtcEndpoint.kmd.json:515]: include -> includes

File: src/server/interface/elements.WebRtcEndpoint.kmd.json [hunk 6]
```diff
@@ -505,6 +505,23 @@
...
+              <p>
+              It must be taken into account that on Internet not all intermediate routers ar guaranteed to enforce those DSCP values, even it is possible that
```
Comment (j1elo) [inline src/server/interface/elements.WebRtcEndpoint.kmd.json:517]: ar -> are

File: src/server/interface/elements.WebRtcEndpoint.kmd.json [hunk 7]
```diff
@@ -914,6 +931,89 @@ This could also happen in the middle of a session, though not likely.
...
+        "AF43",     //  38
+        "EF",       //  46
+        "VOICEADMIT", // 44
+        "LE"        //   1
+      ],
+      "name": "DSCPValue",
+      "doc": "Possible DSCP values 
+      <p>
+      WebRTC recommended values are taken from RFC 8837 https://datatracker.ietf.org/doc/html/rfc8837#section-5 , These are the values from AUDIO_VERYLOW to DATA_HIGH. First element in the 
+      name indicates kind of traffic that it should apply to, the second indicates relative prioriy. For video, a third field would indicate if the traffic is intended for high throughput or not.
```
Comment (j1elo) [inline src/server/interface/elements.WebRtcEndpoint.kmd.json:986]: prioriy -> priority

File: src/server/interface/elements.WebRtcEndpoint.kmd.json [hunk 8]
```diff
@@ -914,6 +931,89 @@ This could also happen in the middle of a session, though not likely.
...
+          +-----------------------+----------+-----+------------+------------+
+          +-----------------------+----------+-----+------------+------------+
+          |          Data         |  LE (1)  |  DF |    AF11    |    AF21    |
+          |                       |          | (0) |            |            |
+          +-----------------------+----------+-----+------------+------------+
+
+      As indicated also in RFC, non interactive video is not considered
+      <p>
+      Apart from the WebRTC recommended values, we also include all possible values are referenced in http://www.iana.org/assignments/dscp-registry/dscp-registry.xml of course some of 
+      those values are synomims for the WebRTC recommended ones, they are included mainly for completeness
```
Comment (j1elo) [inline src/server/interface/elements.WebRtcEndpoint.kmd.json:1011]: synomims -> synonyms

File: src/server/interface/elements.WebRtcEndpoint.kmd.json [hunk 9]
```diff
@@ -914,6 +931,89 @@ This could also happen in the middle of a session, though not likely.
...
+      <p>
+      And as a last point, we include a shorthand for Chrome supported markings for low  (CS0), very low (CS1), medium (CS7) and high (CS7) priorities in priority property for RTCRtpSender parameters. See https://developer.mozilla.org/en-US/docs/Web/API/RTCRtpSender/setParameters
+      <p>
+      This only covers outgoing network packets form KMS, to complete the solution, DSCP must be also requested on client, unfortunatelly for traffic on the other direction, this must be requested to the
```
Comment (j1elo) [inline src/server/interface/elements.WebRtcEndpoint.kmd.json:1015]: form -> from

Comment (j1elo) [inline src/server/interface/elements.WebRtcEndpoint.kmd.json:1015]: unfortunatelly -> unfortunately

**Comment by j1elo** [comment]
Code looks good to me. Sorry for blocking this on the basis of silly minor typos, but there were A LOT of them.

I'd have fixed them myself, normally I ask PRs to have the option "**Allow edits from maintainers**" enabled, so I can make minor changes before merging. But in GitHub that's not possible for PRs created from an Organization's account (as opposed to a private user account). See https://github.com/orgs/community/discussions/5634 -- until they fix their sh*t, I'll have to ask for these minor changes, or otherwise you could create the PRs from a personal user account

**Comment by slabajo** [comment]
Just fixed typos, I am afraid you were right, there were a lot of them, my fault.

**Pull request merged**